Source/Events.Processing/Projections/ProjectionsService.cs

```diff
@@ -0,0 +1,246 @@
+
+                await ResetIfDefinitionChanged(arguments, cancellationToken).ConfigureAwait(false);
+                await eventProcessorStreamProcessor.Initialize().ConfigureAwait(false);
+                return new[] { eventProcessorStreamProcessor.Start(), runningDispatcher };
+            }
+            catch (Exception ex)
+            {
+                if (!cancellationToken.IsCancellationRequested)
+                {
+                    _logger.ErrorWhileStartingProjection(ex, arguments);
+                }
+
+                return ex;
+            }
+        }
+
+        Try<StreamProcessor> TryRegisterProjection(
+            ProjectionRegistrationArguments arguments,
+            Func<IEventProcessor> getEventProcessor,
+            CancellationToken cancellationToken)
+        {
+            _logger.RegisteringProjection(arguments);
+            try
+            {
+                return (_streamProcessors.TryRegister(
+                    arguments.ProjectionDefinition.Scope,
+                    arguments.ProjectionDefinition.Projection.Value,
+                    new EventLogStreamDefinition(),
+                    getEventProcessor,
+                    cancellationToken,
+                    out var outputtedEventProcessorStreamProcessor), outputtedEventProcessorStreamProcessor);
+            }
+            catch (Exception ex)
+            {
+                if (!cancellationToken.IsCancellationRequested)
+                {
+                    _logger.ErrorWhileRegisteringProjection(ex, arguments);
+                }
+
+                return ex;
+            }
+        }
+        async Task ResetIfDefinitionChanged(
+            ProjectionRegistrationArguments arguments,
+            CancellationToken token)
+        {
+            _logger.ComparingProjectionDefintion(arguments);
+            var tenantsAndComparisonResult = await _projectionDefinitionComparer.DiffersFromPersisted(arguments.ProjectionDefinition, token).ConfigureAwait(false);
+
+            await _onAllTenants.PerformAsync(async tenant =>
+            {
+                if (!tenantsAndComparisonResult.TryGetValue(tenant, out var comparisonResult)) return;
+                _logger.PersistingProjectionDefinition(arguments, tenant);
+                if (!await _getProjectionDefinitions().TryPersist(arguments.ProjectionDefinition, token).ConfigureAwait(false)) throw new CouldNotPersistProjectionDefinition(arguments.ProjectionDefinition, tenant);
+                if (!comparisonResult.Succeeded)
+                {
+                    _logger.ResetingProjections(arguments, tenant, comparisonResult.FailureReason);
+                    if (!await _getProjectionStates().TryDrop(arguments.ProjectionDefinition.Projection, arguments.ProjectionDefinition.Scope, token).ConfigureAwait(false)) throw new CouldNotResetProjectionStates(arguments.ProjectionDefinition, tenant);
+                }
```

Comment:
It seems we should do this in the opposite order? If the "new" definition is persisted - but the projections are not dropped. They will have a wrong old state - and not be replayed on next connection right?